### PR TITLE
Feature/unbalanced neural ot

### PR DESCRIPTION
**Pull request by MUCDK**
Implements unbalanced neural OT as described [here](https://www.biorxiv.org/content/10.1101/2022.10.04.510766v1), but an extension to [this](https://arxiv.org/abs/2209.15621) version should be straightforward. 

Moreover, I added a notebook which demonstrates unbalanced OT, both in the discrete and the neural case. Therefore, I also added a new data class.

So far, this is only a draft, as the notebook will have to be rerun.

### CODE REVIEW

File: src/ott/solvers/nn/neuraldual.py [hunk 1]
```diff
@@ -543,6 +573,43 @@ def g_value_finetuned(y: jnp.ndarray) -> jnp.ndarray:
...
+        batch_source,
+        batch_target,
+        epsilon=self.epsilon,
+        scale_cost=self.sample_sinkhorn_kwargs.pop("scale_cost", "mean")
+    )
+    prob = LinearProblem(geom, tau_a=self.tau_a, tau_b=self.tau_b)
+    out = sinkhorn.Sinkhorn(**self.sample_sinkhorn_kwargs)(prob)
+
+    # jax categorical uses log probabilities
+    log_marginals_source = jnp.log(jnp.sum(out.matrix, axis=1))
```
Comment (michalk8) [inline src/ott/solvers/nn/neuraldual.py:602]: Use `jnp.log(out.marginals(1))`

File: src/ott/solvers/nn/neuraldual.py [hunk 2]
```diff
@@ -543,6 +573,43 @@ def g_value_finetuned(y: jnp.ndarray) -> jnp.ndarray:
...
+  def resample(
+      self, batch_source: jnp.ndarray, batch_target: jnp.ndarray,
+      rng: jnp.ndarray
+  ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    r"""Resample data based on posterior marginals of entropy-regularized optimal transport solution.
+
+    Args:
+      batch_source: Batch of the source distribution.
+      batch_target: Batch of the target distribution.
+      rng: :class:`jax.numpy.ndarray` used for setting the seed.
```
Comment (michalk8) [inline src/ott/solvers/nn/neuraldual.py:585]: No need to refer to `jax.numpy` here.

File: src/ott/solvers/nn/neuraldual.py [hunk 3]
```diff
@@ -543,6 +573,43 @@ def g_value_finetuned(y: jnp.ndarray) -> jnp.ndarray:
...
+
+    Returns:
+      Resampled `batch_source` and `batch_target`.
+    """
+    if self.tau_a == 1.0 and self.tau_b == 1.0:
```
Comment (michalk8) [inline src/ott/solvers/nn/neuraldual.py:590]: Consider adding a property called `is_balanced: return self.tau_a == 1.0 and self.tau_b == 1.0`

File: src/ott/solvers/nn/neuraldual.py [hunk 4]
```diff
@@ -543,6 +573,43 @@ def g_value_finetuned(y: jnp.ndarray) -> jnp.ndarray:
         corr=True
     )
 
+  def resample(
+      self, batch_source: jnp.ndarray, batch_target: jnp.ndarray,
+      rng: jnp.ndarray
```
Comment (michalk8) [inline src/ott/solvers/nn/neuraldual.py:578]: jax.random.PRNGKeyArray

File: src/ott/solvers/nn/neuraldual.py [hunk 5]
```diff
@@ -543,6 +573,43 @@ def g_value_finetuned(y: jnp.ndarray) -> jnp.ndarray:
...
+    if self.tau_a == 1.0 and self.tau_b == 1.0:
+      return batch_source, batch_target
+    geom = PointCloud(
+        batch_source,
+        batch_target,
+        epsilon=self.epsilon,
+        scale_cost=self.sample_sinkhorn_kwargs.pop("scale_cost", "mean")
+    )
+    prob = LinearProblem(geom, tau_a=self.tau_a, tau_b=self.tau_b)
+    out = sinkhorn.Sinkhorn(**self.sample_sinkhorn_kwargs)(prob)
```
Comment (michalk8) [inline src/ott/solvers/nn/neuraldual.py:599]: For simplicity, I'd use `sinkhorn.solve(geom, tau_a=self.tau_a, tau_b=self.tau_b, **self.solve_kwargs)` instead of instantiating the problem and the solver.

File: src/ott/solvers/nn/neuraldual.py [hunk 6]
```diff
@@ -97,6 +101,12 @@ class W2NeuralDual:
     parallel_updates: Update :math:`f` and :math:`g` at the same time
     init_f_params: initial parameters for :math:`f`
     init_g_params: initial parameters for :math:`g`
+    tau_a: If `< 1`, defines how much unbalanced the problem is
```
Comment (michalk8) [inline src/ott/solvers/nn/neuraldual.py:104]: In the above documentation, right after `:attr:~ott.solvers.nn.models.ModelBase.is_potential.`, I'd mention your approach for unbalancedness + cite the relevant papers.

**Review by michalk8 — CHANGES_REQUESTED** [review]
I reviewed only the code, also please run pre-commit.

File: docs/tutorials/notebooks/unbalanced_optimal_transport.ipynb [hunk 7]
```diff
@@ -0,0 +1,726 @@
+{
```
Comment (bamos) [inline docs/tutorials/notebooks/unbalanced_optimal_transport.ipynb:1]: <p>This looks like it really blew up, were you able to fix it?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/289/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

File: src/ott/solvers/nn/neuraldual.py (hunk 5 above)
Comment (MUCDK) [reply src/ott/solvers/nn/neuraldual.py:599]: I thought the functional API was going to be removed?

File: src/ott/solvers/nn/neuraldual.py [hunk 8]
```diff
@@ -97,6 +101,12 @@ class W2NeuralDual:
...
+      on the first marginal.
+    tau_b: If `< 1`, defines how much unbalanced the problem is
+      on the second marginal.
+    epsilon: regularisation parameter in the inner sinkhorn loop. Only relevant, if `tau_a!=1` or `tau_b!=1`.
```
[the remark this answers is not in the document]
Comment (MUCDK) [reply src/ott/solvers/nn/neuraldual.py:108]: depends on whether we go with functional or non-functional interface now, right?

File: src/ott/solvers/nn/neuraldual.py (hunk 1 above)
Comment (MUCDK) [reply src/ott/solvers/nn/neuraldual.py:602]: Sorry, I couldn't find `marginals` to be an attribute of the `SinkhornOutput`, could you direct me to that please?

File: src/ott/solvers/nn/neuraldual.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (michalk8) [reply src/ott/solvers/nn/neuraldual.py:108]: This should be independent of it, since for both APIs, you need to construct a geometry first.

File: src/ott/solvers/nn/neuraldual.py (hunk 1 above)
Comment (michalk8) [reply src/ott/solvers/nn/neuraldual.py:602]: Sorry, there was a typo, it's `out.marginal(1)`, see [here](https://github.com/ott-jax/ott/blob/main/src/ott/solvers/linear/sinkhorn.py#L399).

File: docs/tutorials/notebooks/unbalanced_optimal_transport.ipynb (hunk 7 above)
Comment (MUCDK) [reply docs/tutorials/notebooks/unbalanced_optimal_transport.ipynb:1]: <p>Unfortunately not yet, which is a bit surprising as training was really stable for our paper. In fact, there are two  difference, which is that we trained with early stopping and best model performance - and the <code>amortization_loss</code></p><p>Then I tried running it with <code>amortization_loss=regression</code> (the fully balanced case, so code changes should not have changed, but I can check again on the original branch), and it threw an error. </p><pre class="ql-syntax" spellcheck="false">TypeError                                 Traceback (most recent call last)
Cell In[16], line 41
     25         plt.close(fig)
     28 neural_dual_solver = neuraldual.W2NeuralDual(
     29     input_dim,
     30     neural_f,
   (...)
     39     #epsilon=1,
     40 )
---&gt; 41 learned_potentials = neural_dual_solver(
     42     *train_dataloaders,
     43     *valid_dataloaders,
     44     callback=training_callback,
     45 )
     46 clear_output()

File ~/git_repos/ott-1/src/ott/solvers/nn/neuraldual.py:255, in W2NeuralDual.__call__(self, trainloader_source, trainloader_target, validloader_source, validloader_target, callback)
    246 def __call__(
    247     self,
    248     trainloader_source: Iterable[jnp.ndarray],
   (...)
    253 ) -&gt; Union[potentials.DualPotentials, Tuple[potentials.DualPotentials,
    254                                             Train_t]]:
--&gt; 255   logs = self.train_fn(
    256       trainloader_source,
    257       trainloader_target,
    258       validloader_source,
    259       validloader_target,
    260       callback=callback,
    261   )
    262   res = self.to_dual_potentials()
    264   return (res, logs) if self.logging else res

File ~/git_repos/ott-1/src/ott/solvers/nn/neuraldual.py:377, in W2NeuralDual.train_neuraldual_alternating(self, trainloader_source, trainloader_target, validloader_source, validloader_target, callback)
    374   batch_g["source"] = jnp.asarray(next(trainloader_source))
    375   batch_g["target"] = jnp.asarray(next(trainloader_target))
--&gt; 377   self.state_g, loss_g, _ = self.train_step_g(
    378       self.state_f, self.state_g, batch_g
    379   )
    381 # get train batch for potential f
    382 batch_f["source"] = jnp.asarray(next(trainloader_source))

    [... skipping hidden 12 frame]

File ~/git_repos/ott-1/src/ott/solvers/nn/neuraldual.py:502, in W2NeuralDual.get_step_fn.&lt;locals&gt;.step_fn(state_f, state_g, batch)
    499 grad_fn = jax.value_and_grad(loss_fn, argnums=[0, 1], has_aux=True)
    500 if train:
    501   # compute loss and gradients
--&gt; 502   (loss, (loss_f, loss_g, W2_dist)), (grads_f, grads_g) = grad_fn(
    503       state_f.params,
    504       state_g.params,
    505       state_f.potential_value_fn,
    506       state_g.potential_value_fn,
    507       state_g.potential_gradient_fn,
    508       batch,
    509   )
    511   # update state
    512   if to_optimize == "both":

    [... skipping hidden 8 frame]

File ~/git_repos/ott-1/src/ott/solvers/nn/neuraldual.py:464, in W2NeuralDual.get_step_fn.&lt;locals&gt;.loss_fn(params_f, params_g, f_value, g_value, g_gradient, batch)
    459 elif self.amortization_loss == 'objective':
    460   f_value_parameters_detached = lambda x: f_value(
    461       jax.lax.stop_gradient(params_f), x
    462   )
    463   amor_loss = (
--&gt; 464       f_value_parameters_detached(init_source_hat) -
    465       batch_dot(init_source_hat, target)
    466   ).mean()
    467 else:
    468   raise ValueError("Amortization loss has been misspecified.")

TypeError: unsupported operand type(s) for -: 'function' and 'JVPTracer' 
</pre>

Comment (bamos) [reply docs/tutorials/notebooks/unbalanced_optimal_transport.ipynb:1]: Ah sorry about that, maybe we should update the neural dual test to sweep over `amortization_loss` in "regression" and "objective". That error can be fixed by updating that line of `neuraldual` to:

```python
elif self.amortization_loss == 'objective':
  f_value_parameters_detached = f_value(
      jax.lax.stop_gradient(params_f), g_value_partial)
  amor_loss = (
      f_value_parameters_detached(init_source_hat) -
      batch_dot(init_source_hat, target)
  ).mean()
```

Can you push this update into this PR? I could also make another small PR with this but it seems a little simpler to keep it in here

Comment (bamos) [reply docs/tutorials/notebooks/unbalanced_optimal_transport.ipynb:1]: With the objective loss your example still seems to be blowing up. I think it could be related to the ICNN architecture as I wasn't able to find an extremely stable/optimal one. If you have another architecture from your paper it may be worth trying that instead of the default OTT one. I've also found the MLP potentials to work well and be stable in a lot of cases.

image

Comment (MUCDK) [reply docs/tutorials/notebooks/unbalanced_optimal_transport.ipynb:1]: Oh I think I did not use the objective loss here, sorry, the order of execution is not yet correct, as I just did this push now to show the small bug. Thanks for letting me know how to fix it.

We actually used the same architecture, i.e. with [64,64,64,64] layers. I'll give it another try tomorrow!

Comment (bamos) [reply docs/tutorials/notebooks/unbalanced_optimal_transport.ipynb:1]: Maybe we should make this clearer somewhere but you can also disable the conjugate fine-tuning by setting `conjugate_solver=None` in case that's messing up the convergence

**Comment by bamos** [comment]
Continuing on [this thread](https://github.com/ott-jax/ott/pull/289#discussion_r1107313344), I've created [this PR](https://github.com/MUCDK/ott-1/pull/1) into the fork that fixes a few bugs in the neuraldual (when using objective-based amortization and no fine-tuning) and have added these cases to the tests. @MUCDK, can you merge that in to your fork/branch for this PR, and then we can continue discussing changes in here? (It also looks like you'll have to merge the upstream main into your fork and fix a conflict after merging my PR)

I've also found a better starting point for the neural dual solver that uses an ICNN for the `f` potential with an MLP for the conjugate's gradient (with default values close to the ones used in the main neural dual tutorial). The ICNN on the balanced problem before was diverging to:

image

The updates in my PR at least bump it up to:

image

And then setting the unbalanced mode gets close to the expected solution:

image

These still aren't perfect so there's still some tuning to be done, but I hope this helps as a better starting point! (And it'll let you try out the options with the objective amortization loss and no fine-tuning if you want to try those) You probably know more about debugging this setting than I do, but I'm optimistic it could be finalized by running the training for a little longer than (5-10k updates), adding a LR decay to the optimizers, and/or updating the ICNN architecture/initialization

**Comment by bamos** [comment]
I also think that @theouscidda6 has some other updates to the ICNN architecture to make it work better on the W2 benchmark that may also fix the usage of the ICNN here

**Comment by lucaeyring** [comment]
Hey @bamos, thanks for your work on Neural OT and also helping with this PR! I've merged your fixes into the PR.
I've been working on a small revamp of this PR to reproduce the results of our paper. I managed to do this using a potential MLP for both $f$ and $g$ and the default ```conjugate_solver``` and ```amortization_loss```. This setting also worked well for the toy experiments in ```neural_dual.ipynb```.

Balanced:
Bildschirmfoto 2023-03-02 um 13 33 07 Bildschirmfoto 2023-03-02 um 13 33 25   Bildschirmfoto 2023-03-02 um 13 33 14

Unbalanced:

Bildschirmfoto 2023-03-02 um 13 51 23 Bildschirmfoto 2023-03-02 um 13 51 28 Bildschirmfoto 2023-03-02 um 13 51 16


However, I was not able to get a similarly clean result using the current ICNN architecture. When removing the ```PosDefPotentials``` layer and leveraging identity pretraining instead of the analytic initialization, I'm able to reproduce the results with both a conjugate solver, as well as with the alternating Makkuva training. I didn't want to change the architecture though, so in the notebook, I'm now using MLPs.

To me, it seems like the ```PosDefPotentials``` layer might be causing problems here though. I saw that in your work, for ICNNs you were also not using a ```PosDefPotentials``` layer and relying on pretraining, did you discuss adding this option to ott-jax?

In addition to that, I also added a few general things:
- Jitted sampling in ```dataset.py```, this significantly speeds up training
- When using alternating training, I believe it is not necessary to sample an extra training batch for the update of ```f```. Instead one can use the last training batch used to update ```g```. 
- The option to plot the displacement potentials $\dfrac{1}{2}||\cdot||^2_2 - f(\cdot)$ instead of the convex potential $f(\cdot)$. Above I'm also plotting these. They should be more interpretable, see e.g.:
figure_displ_potential

@michalk8 would be great if you could review this. Regarding the displacement potential, I could also add some explanatory examples/documentation into e.g. the ```neural_dual.ipynb```.

**Comment by bamos** [comment]
That's great to see @lucaeyring!

> To me, it seems like the `PosDefPotentials` layer might be causing problems here though. I saw that in your work, for ICNNs you were also not using a `PosDefPotentials` layer and relying on pretraining, did you discuss adding this option to ott-jax?

That's right, the neural potentials in the [w2ot](https://github.com/facebookresearch/w2ot) repo are not using `PosDefPotentials` and are using identity pre-training. The other difference there is that the `PositiveDense` layer has an additional scaling term [here](https://github.com/facebookresearch/w2ot/blob/main/w2ot/external/ott_icnn.py#L57). We didn't discuss adding pre-training to `ott-jax`, but it could make a lot of sense to support a pre-training phase for: 1) identity initialization, 2) variations on the Gaussian initialization (e.g., only one of the measures is approximated with a Gaussian but the other is not), 3) fitting the neural potentials to sinkhorn/entropic solutions on a smaller batch

> * Jitted sampling in `dataset.py`, this significantly speeds up training
> * When using alternating training, I believe it is not necessary to sample an extra training batch for the update of `f`. Instead one can use the last training batch used to update `g`.
> * The option to plot the displacement potentials 12||⋅||22−f(⋅) instead of the convex potential f(⋅). Above I'm also plotting these.

These all make sense too

File: src/ott/solvers/nn/neuraldual.py [hunk 9]
```diff
@@ -245,7 +313,7 @@ def __call__(  # noqa: D102
         validloader_target,
         callback=callback,
     )
-    res = self.to_dual_potentials()
+    res = self.to_dual_potentials(finetune_g=self.conjugate_solver is not None)
```
Comment (michalk8) [inline src/ott/solvers/nn/neuraldual.py:316]: This is no longer needed thanks to #327, please remove (also from all the other calls).

File: src/ott/solvers/nn/neuraldual.py [hunk 10]
```diff
@@ -200,6 +230,44 @@ def setup(
...
+    @jax.jit
+    def _unbalanced_resample(
+        key: jax.random.KeyArray,
+        batch: jnp.ndarray,
+        log_marginals: jnp.ndarray,
+    ) -> jnp.ndarray:
+      """Jitted resample function of a batch based upon log marginals."""
+      # sample from marginals
+      indices = jax.random.categorical(
+          key, log_marginals, shape=[batch.shape[0]]
```
Comment (michalk8) [inline src/ott/solvers/nn/neuraldual.py:264]: `shape=batch.shape[:1]` should also work

File: docs/problems/nn.rst [hunk 11]
```diff
@@ -11,5 +11,6 @@ Dataset
     :toctree: _autosummary
 
     dataset.create_gaussian_mixture_samplers
```
Comment (michalk8) [inline docs/problems/nn.rst:13]: I think I'd prefer to have shorter names, like `gaussian_mixture_sampler(s)`, can you please adjust (will need to also change the other NN notebooks).

**Review by michalk8 — CHANGES_REQUESTED** [review]
Regarding docs linter:
- `unbalancedness` can be added to `docs/spelling/technical.txt`
- sinkhorn -> Sinkhorn
- the 403 for https://onlinelibrary.wiley.com/doi/10.1002/mana.19901470121 - if it pops up again, please add to `linkcheck_ignore` in `docs/conf.py`

Will review the NB shortly.

**Comment by lucaeyring** [comment]
<p>We noticed that depending on the data, sinkhorn did not converge with the default <code>max_iterations </code>so we increased this. Since it worked well with higher <code>max_iterations</code> we didn't benchmark much further. I now changed this to instead use <code>recenter_potentials=True</code> and this also seems to work well. </p>

---

 <a href='https://app.reviewnb.com/ott-jax/ott/pull/289/discussion/#comment-1453413364'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by lucaeyring** [comment]
Hey @michalk8 , I have now implemented your comments. I removed the unbalanced discrete OT part in the notebook and renamed it to `unbalanced_neural_dual.ipynb`. Additionally, I adjusted plotting of the data in both `neural_dual` notebooks to now use `ott.tools.plot`. Lastly, I added a short explanation of the displacement potential at the end of the `neural_dual.ipynb`.

File: src/ott/problems/linear/potentials.py [hunk 12]
```diff
@@ -310,9 +313,11 @@ def plot_potential(
     X1, X2 = jnp.meshgrid(x1, x2)
     X12flat = jnp.hstack((X1.reshape(-1, 1), X2.reshape(-1, 1)))
     Zflat = jax.vmap(self.f if forward else self.g)(X12flat)
+    if displacement:
+      Zflat = 0.5 * (jnp.linalg.norm(X12flat, axis=-1) ** 2) - Zflat
     Zflat = np.asarray(Zflat)
-    vmin, vmax = np.quantile(Zflat, [quantile, 1. - quantile])
```
Comment (michalk8) [inline src/ott/problems/linear/potentials.py:314]: Why was this removed?

File: tests/solvers/nn/neuraldual_test.py [hunk 13]
```diff
@@ -55,6 +55,8 @@ class TestNeuralDual:
       back_and_forth=[True, False],
       amortization_loss=["objective", "regression"],
       conjugate_solver=[conjugate_solvers.DEFAULT_CONJUGATE_SOLVER, None],
+      tau_a=[0.8, 1.0],
```
Comment (michalk8) [inline tests/solvers/nn/neuraldual_test.py:58]: I think it's better to have a separate test for this, as previously the CI was failing because it took so long...

File: src/ott/solvers/nn/neuraldual.py [hunk 14]
```diff
@@ -347,20 +416,20 @@ def train_neuraldual_alternating(
...
+          rng_train, rng = jax.random.split(rng, 2)
+          train_batch = self.unbalanced_resample(rng_train, train_batch)
+        # update g
         self.state_g, loss_g, _ = self.train_step_g(
-            self.state_f, self.state_g, batch_g
+            self.state_f, self.state_g, train_batch
         )
-
-      # get train batch for potential f
-      batch_f["source"] = jnp.asarray(next(trainloader_source))
```
Comment (michalk8) [inline src/ott/solvers/nn/neuraldual.py:359]: Why isn't a new batch sampled from the `trainloader_{source,target}` anymore?

File: docs/problems/nn.rst [hunk 15]
```diff
@@ -10,6 +10,7 @@ Dataset
 .. autosummary::
     :toctree: _autosummary
 
-    dataset.create_gaussian_mixture_samplers
+    dataset.gaussian_mixture_samplers
+    dataset.uniform_mixture_samplers
     dataset.Dataset
     dataset.GaussianMixture
```
Comment (michalk8) [inline docs/problems/nn.rst:16]: I think this can be removed from here, users should interact with the `{gaussian,uniform}_mixture_samplers` only.

File: src/ott/problems/nn/dataset.py [hunk 16]
```diff
@@ -85,27 +89,139 @@ def __post_init__(self):
...
+  return train_dataset, valid_dataset, dim_data
+
+
+@dataclasses.dataclass
+class UniformMixture:
+  """A mixture of uniform distributions.
+
+  Args:
+    name: the name specifying position of mixture of Gaussian:
+        - ``'top'`` (uniform distributions on top),
```
Comment (michalk8) [inline src/ott/problems/nn/dataset.py:164]: Missing empty row after `Gaussian:`.
Also please dedent the list as
```rst
name: ... mixture of Gaussians:

  - ``'top'``
  - ...
```

File: src/ott/problems/nn/dataset.py [hunk 17]
```diff
@@ -85,27 +89,139 @@ def __post_init__(self):
...
+
+
+def gaussian_mixture_samplers(
+    name_source: Arrangement_t,
+    name_target: Arrangement_t,
+    train_batch_size: int = 2048,
+    valid_batch_size: int = 2048,
+    rng: jax.random.PRNGKeyArray = jax.random.PRNGKey(0),
+) -> Tuple[Dataset, Dataset, int]:
+  """Gaussian samplers for :class:`~ott.solvers.nn.neuraldual.W2NeuralDual`.
```
Comment (michalk8) [inline src/ott/problems/nn/dataset.py:125]: Should be `Uniform samplers for ...`, right?

**Review by michalk8 — CHANGES_REQUESTED** [review]

**Comment by michalk8** [comment]
Pinging @bunnech - would you also like to review this PR and/or do you have any comments?

File: src/ott/problems/linear/potentials.py (hunk 12 above)
Comment (lucaeyring) [reply src/ott/problems/linear/potentials.py:314]: I'm sorry I forgot to clarify this. When plotting the displacement potential, the target distribution will usually be represented by the minimum of the potential. Removing the minimum 5 percent of this flattens out these minima, which in my opinion is undesirable.

File: tests/solvers/nn/neuraldual_test.py (hunk 13 above)
Comment (lucaeyring) [reply tests/solvers/nn/neuraldual_test.py:58]: What should the separate test look like? Just less tested options so it doesn't take as long? Should I just remove this or also add the new unbalanced neural dual test?

File: src/ott/problems/linear/potentials.py (hunk 12 above)
Comment (lucaeyring) [reply src/ott/problems/linear/potentials.py:314]: Removed it for now.

File: tests/solvers/nn/neuraldual_test.py (hunk 13 above)
Comment (lucaeyring) [reply tests/solvers/nn/neuraldual_test.py]: Removed it for now.
